src/neo4j_genai/retrievers/external/weaviate/weaviate.py

```diff
@@ -0,0 +1,138 @@
+from neo4j_genai.retrievers.utils import validate_search_query_input
+import weaviate.classes as wvc
+import neo4j
+import logging
+from neo4j_genai.neo4j_queries import get_search_query
+
+logger = logging.getLogger(__name__)
+
+
+class WeaviateNeo4jRetriever(ExternalRetriever):
+    def __init__(
+        self,
+        driver: neo4j.Driver,
+        client: any,
+        collection: str,
+        id_property_external: str,
+        id_property_neo4j: str = None,
+        embedder: Optional[Embedder] = None,
+        return_properties: Optional[list[str]] = None,
+        retrieval_query: Optional[str] = None,
+    ):
+        self.driver = driver
+        self.client = client
+        self.search_collection = client.collections.get(collection)
+        self._id_property_external = id_property_external
+        self._id_property_neo4j = id_property_neo4j
+        self.embedder = embedder
+        self.return_properties = return_properties
+        self.retrieval_query = retrieval_query
+
+    @property
+    def id_property_external(self):
+        return self._id_property_external
+
+    @property
+    def id_property_neo4j(self):
+        return self._id_property_neo4j
+
+    def search(
+        self,
+        query_vector: Optional[list[float]] = None,
+        query_text: Optional[str] = None,
+        top_k: int = 5,
+    ) -> list[neo4j.Record]:
+        """Get the top_k nearest neighbor embeddings using Weaviate for either provided query_vector or query_text.
+        Both query_vector and query_text can be provided.
+        If query_vector is provided, then it will be preferred over the embedded query_text
+        for the vector search.
+        If query_text is provided, then it will check if an embedder is provided and use it to generate the query_vector.
+        If no embedder is provided, then it will assume that the vectorizer is used in Weaviate.
+
+        See the following documentation for more details:
+        - [Query a vector index](https://neo4j.com/docs/cypher-manual/current/indexes-for-vector-search/#indexes-vector-query)
+        - [db.index.vector.queryNodes()](https://neo4j.com/docs/operations-manual/5/reference/procedures/#procedure_db_index_vector_queryNodes)
+        - [db.index.fulltext.queryNodes()](https://neo4j.com/docs/operations-manual/5/reference/procedures/#procedure_db_index_fulltext_querynodes)
+        Args:
+            query_text (str): The text to get the closest neighbors of.
+            query_vector (Optional[list[float]], optional): The vector embeddings to get the closest neighbors of. Defaults to None.
+            top_k (int, optional): The number of neighbors to return. Defaults to 5.
+        Raises:
+            ValueError: If validation of the input arguments fail.
+        Returns:
+            list[neo4j.Record]: The results of the search query
+        """
+
+        validate_search_query_input(query_text=query_text, query_vector=query_vector)
```

Comment:
+1 on that, it will probably look the same as `VectorSearchModel`.

In fact after reviewing the "search model" types, we could refactor to only have two types: `SearchModel` and `CypherSearchModel`, since we already do validation of the index names in a separate data model in the `init`. We can work on this later and doesn't have to block this PR

Reply:
That way we only do data validation for the input arguments into the `.search()` method